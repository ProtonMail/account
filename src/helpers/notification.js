import appDispatcher from 'frontend-commons/src/utils/appDispatcher';

const dispatcher = appDispatcher('notification');

const action = (type) => (message, opt = {}) => dispatcher.input(type, { message, opt });
export const success = action('success');
export const error = (err) => action('error')(err.message, err);
export const info = action('info');

export const onInput = dispatcher.onInput;
