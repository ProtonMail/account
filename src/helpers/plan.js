const searchParams = new URLSearchParams(window.location.search);

export const getDefaultConfigPlan = ({ Cycle = 12, Currency = 'EUR' } = {}) => ({
  Currency: searchParams.get('currency') || Currency,
  Cycle: +(searchParams.get('billing') || Cycle),
  Name: searchParams.get('plan') || 'vpnplus'
});

export const cycleString = ({ Cycle }, isMonth = false) => (Cycle === 12 && !isMonth) ? 'year' : 'month';
