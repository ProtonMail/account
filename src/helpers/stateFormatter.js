/**
 * Format the state and extend it as it's not recursive with unistore
 * @param  {Object} state
 * @param  {String} key
 * @param  {Object} value
 * @return {Object}       new state
 * @function
 */
export const toState = (state, key, value) => ({ [key]: { ...state[key], ...value } });

/**
 * Format the state and extends using the dot notation.
 * @param {Object} state - the state to update
 * @param {string|string[]} extendedKey - the key to update in the state. It can be either a string, either a list of string.
 * If it's a string, it corresponds to a path, the `.` separates the keys. If it's an array, each element is a key.
 * @param {Object} value - the updated value.
 * @returns {Object} new state.
 * @example
 * > extended(
 *     {
 *         a: 1,
 *         b: {
 *             c: {
 *                 d: 2,
 *                 e: 3
 *             },
 *             f: {
 *                 g: 4
 *             }
 *         }
 *     },
 *     'b.c',
 *     { d: 42, f: 25 }
 * );
 * // returns
 * {
 *     // a removed, because root keys are handled by unistore
 *     b: {
 *         c: {
 *             d: 42, // updated
 *             f: 25, // new
 *             e: 3   // unchanged
 *         },
 *         f: { // f is not removed, because subkeys are not handled by unistore
 *             g: 4
 *         }
 *     }
 * }
 *
 * @example
 * // beware!
 * > extended({...sameAsBefore }, ['b.c'], { d: 42, f: 25 })
 * { 'b.c': { d: 42, f: 25 } } // the new key 'b.c' is created
 * @function
 */
export const extended = (state, extendedKey, value) => {
    if (!Array.isArray(extendedKey)) {
        return extended(state, extendedKey.split('.'), value);
    }
    if (!extendedKey.length) {
        return value;
    }

    const [firstKey, ...rest] = extendedKey;
    return toState(state, firstKey, extended(state[firstKey], rest, value));
};
