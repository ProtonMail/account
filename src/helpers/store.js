import { createStore } from 'unistore/full/preact';
import devtools from 'unistore/devtools';

import authActions from '../actions/authentication';
import settingsActions from '../actions/settings';
import scopeActions from '../actions/scope';

export const initialState = {
    auth: {
        isLoggedIn: false,
        user: {},
        step: 'login',
        twoFactorResponse: {}
    },
    scope: {
        creds: null,
        response: null
    },
    settings: {
        addU2FKey: {},
        reset2FARecoveryCodes: {}
    }
};
const store = createStore(initialState);


export const actions = (store) => ({
    ...authActions(store),
    ...settingsActions(store),
    ...scopeActions(store)
});

export default store;
