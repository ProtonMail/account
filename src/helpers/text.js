import { saveAs } from 'file-saver';

/**
 * Download the given lines as a file.
 * @param {string} name - the name of the file
 * @param {string[]} lines - the lines to put in the file
 * @param {string} [type='text/plain;charset=utf-8'] - the mime type of the file
 */
export const downloadAsFile = (name, lines = [], type = 'text/plain;charset=utf-8') => {
    const blob = new Blob([lines.join('\r\n')], { type });
    saveAs(blob, name);
};
