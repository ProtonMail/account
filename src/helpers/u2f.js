import { sign, register } from 'u2f-api';
import appProvider from 'frontend-commons/src/appProvider';

export const ERROR_CODE = {
    SUCCESS: 0,
    OTHER_ERROR: 1,
    BAD_REQUEST: 2,
    CONFIGURATION_UNSUPPORTED: 3,
    DEVICE_INELIGIBLE: 4,
    TIMEOUT: 5
};

const ERROR_MAP = {
    [ERROR_CODE.SUCCESS]: () => {
        throw new Error('This is a success!');
    },
    [ERROR_CODE.OTHER_ERROR]: () => 'An error occurred',
    [ERROR_CODE.BAD_REQUEST]: () => 'An internal error occurred.',
    [ERROR_CODE.CONFIGURATION_UNSUPPORTED]: () => 'This security key is not supported.',
    [ERROR_CODE.DEVICE_INELIGIBLE]: (register) =>
        register ? 'This security key is already registered for your account!' : 'This security key is not recognized.',
    [ERROR_CODE.TIMEOUT]: () =>
        'Looks like you are taking too long to respond, please try again with a bit of motivation.'
};

/**
 * Sends a SIGN request to the U2F device
 * @param {Object} U2FRequest - the U2FRequest
 * @param {Object[]} U2FRequest.RegisteredKeys - the registered keys
 * @param {String} U2FRequest.Challenge - the challenge
 * @return {Promise<SignResponse>}
 */
export async function signU2F({ RegisteredKeys: registeredKeys, Challenge: challenge }) {
    const { appId, timeout } = appProvider.getConfig('u2f');

    const signRequest = registeredKeys.map(({ Version: version, KeyHandle: keyHandle }) => ({
        version,
        keyHandle,
        appId,
        challenge
    }));

    return await sign(signRequest, timeout);
}

/**
 * Sends a REGISTER request to the U2F device.
 * @param {Object} U2FRequest - the U2FRequest
 * @param {Object[]} U2FRequest.RegisteredKeys - the registered keys
 * @param {String} U2FRequest.Challenge - the challenge
 * @param {String[]} U2FRequest.Versions - the different versions accepted by the server
 * @return {Promise<RegisterResponse>}
 */
export async function registerU2F({ RegisteredKeys: registeredKeys, Challenge: challenge, Versions: versions }) {
    const { appId, timeout } = appProvider.getConfig('u2f');
    const signRequest = registeredKeys.map(({ Version: version, KeyHandle: keyHandle }) => ({
        version,
        keyHandle
    }));

    const registerRequests = versions.map((version) => ({
        version,
        challenge,
        appId
    }));

    return await register(registerRequests, signRequest, timeout);
}

/**
 * Computes an error message from a U2F error code.
 * @param {number} errorCode - an error code.
 * @param register
 * @return {*}
 */
export function getErrorMessage(errorCode, register = false) {
    if (ERROR_MAP[errorCode]) {
        return ERROR_MAP[errorCode](register);
    }
    throw new Error(`The error code "${errorCode}" does not exist`);
}
