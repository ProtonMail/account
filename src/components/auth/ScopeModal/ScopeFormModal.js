import { Component } from 'preact';
import { Content as ModalContent, Footer as ModalFooter, Wrapper as ModalWrapper } from '../../ui/Modal/index';
import { connect } from 'unistore/full/preact';

import scopeActions from '../../../actions/scope';
import TextButton from '../../ui/TextButton';

import style from './index.css';
import { getErrorMessage } from '../../../helpers/u2f';

/**
 * Modal Form to asks credentials information (password and 2FA).
 */
class ScopeFormModal extends Component {

    componentWillMount() {
        if (this.props.scope.used) {
            this.props.skip();
        }
    }

    componentDidMount() {
        this.props.unscopeInitAction();
        this.setState({ data: { ...this.props.scope.creds } });
    }

    constructor(props) {
        super(props);

        this.state = {
            data: { ...props.scope.creds }
        };
    }

    /**
     * Stores the data is the correct field.
     * @param {Event} e - the event to handle
     * @param {string} e.target.name - the name of the event to handle.
     * @param {string} e.target.value - the value to handle.
     */
    onFieldUpdated({ target: { name, value } }) {
        this.setState({ data: { ...this.state.data, [name]: value } });
    }

    /**
     * Renders the 2FA fields, if needed.
     * @returns {Component}
     */
    renderTwoFactor() {
        const info = this.props.scope.response;
        if (!info) {
            return <div class={style.scopeFormModal}>Requesting authorization...</div>;
        }
        if (!info.TwoFactor) {
            // returning empty div, to keep the place
            return <div class={style.scopeFormModal}/>;
        }

        const {
            creds: { U2F } = {},
            U2FRequest: { status, error } = {}
        } = this.props.scope;

        if (U2F) {
            return (<div class={style.scopeFormModal}>
                <span>Your security key was used. </span>
                <TextButton onClick={this.props.unscopeResetTwoFactorAction}>
                    Undo
                </TextButton>
            </div>);
        }

        if (status === 'pending') {
            return (<div class={style.scopeFormModal}>
                <p>
                    <span>Please activate your security key... </span>
                    <TextButton onClick={this.props.unscopeResetTwoFactorAction}>
                        Cancel
                    </TextButton></p>
            </div>);
        }

        if (status === 'failure') {
            return (<div class={style.scopeFormModal}><p>
                <span>{getErrorMessage(error)} You can </span>
                <TextButton onClick={this.props.unscopeU2FAction}>
                    Try again
                </TextButton>
                <span> or </span>
                <TextButton onClick={this.props.unscopeResetTwoFactorAction}>
                    Use a code
                </TextButton>
                <span>.</span>
            </p>
            </div>);
        }

        const components = [<div class={[style.scopeFormModal, 'form-row'].join(' ')}>
            <label htmlFor="twoFactorCode" class={style.label}>2FA code</label>
            <div>
                <input
                    class={style.input}
                    type="text"
                    name="twoFactorCode"
                    id="twoFactorCode"
                    autoCapitalize="off"
                    autoCorrect="off"
                    autoComplete="off"
                    minLength="6"
                    maxLength="8"
                    required
                    value={this.state.data.twoFactorCode}
                    onInput={(e) => this.onFieldUpdated(e)}/>
            </div>
        </div>];
        if (info['2FA'].U2F) {
            components.push((<div class={style.scopeFormModal}>
                <p>
                    <TextButton onClick={() => this.props.unscopeU2FAction()}>
                        Or use your security key
                    </TextButton>
                </p>
            </div>));
        }
        return components;
    }

    render() {
        const scope = this.props.scope;
        return (
            <ModalWrapper
                onSubmit={(e) => {
                    e.preventDefault();
                    this.props.unscopePasswordAction(this.state.data);
                    this.props.onSubmit();
                }}
                onReset={(e) => {
                    e.preventDefault();
                    this.props.onCancel();
                }}
            >
                <ModalContent>
                    {!!this.props.message && (
                        <div>{this.props.message}</div>
                    )}
                    <div class={[style.scopeFormModal, 'form-row'].join(' ')}>
                        <label htmlFor="password" class={style.label}>password</label>
                        <div>
                            <input
                                class={style.input}
                                type="password"
                                name="password"
                                id="password"
                                value={this.state.data.password}
                                required
                                placeholder="Password"
                                onInput={(e) => this.onFieldUpdated(e)}
                                autoFocus={true}
                            />
                        </div>
                    </div>
                    {this.renderTwoFactor()}
                </ModalContent>
                <ModalFooter>
                    <button type="reset" value="Reset">
                        Back
                    </button>
                    <button type="submit" value="Submit" disabled={!scope.response}>
                        Next
                    </button>
                </ModalFooter>
            </ModalWrapper>
        );
    }
}

export default connect(['scope', 'auth'], scopeActions)(ScopeFormModal);
