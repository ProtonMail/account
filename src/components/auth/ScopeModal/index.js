import ScopeFormModal from './ScopeFormModal';
import store from '../../../helpers/store';
import actions from '../../../actions/scope';

export const steps = (scope) => {
    if (scope !== 'password') {
        return [];
    }
    return [
        {
            title: 'Confirm your identity',
            component: ({ onNextStep, onPreviousStep, onSkipStep, message }) => (
                <ScopeFormModal
                    onSubmit={onNextStep}
                    onCancel={onPreviousStep}
                    skip={onSkipStep}
                    message={message}/>)
        }
    ];
};

export const beforeDismiss = () => {
    const { resetScopeStateAction } = actions(store);
    resetScopeStateAction(store.getState());
};
