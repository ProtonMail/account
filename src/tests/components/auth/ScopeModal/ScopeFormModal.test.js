import render from 'preact-render-to-string';
import { deep } from 'preact-render-spy';

import { ScopeFormModal } from '../../../../components/auth/ScopeModal/ScopeFormModal';
import { ERROR_CODE } from '../../../../helpers/u2f';

describe('ScopeModal step ScopeFormModal', () => {
    describe('initial display', () => {
        test('requesting authorization', () => {
            expect(render(<ScopeFormModal scope={{}}/>)).toMatchSnapshot();
        });

        test('with a message', () => {
            expect(render(<ScopeFormModal
                message="This is a simple message"
                scope={{ response: {} }}/>)
            ).toMatchSnapshot();
        });

        test('without 2FA', () => {
            expect(render(<ScopeFormModal scope={{ response: {} }}/>)).toMatchSnapshot();
        });

        test('with TOTP only', () => {
            const scope = {
                response: {
                    TwoFactor: 1,
                    '2FA': {
                        TOTP: 1,
                        U2F: null
                    }
                }
            };

            expect(render(<ScopeFormModal scope={scope}/>)).toMatchSnapshot();
        });

        test('with U2F only', () => {
            const scope = {
                response: {
                    TwoFactor: 1,
                    '2FA': {
                        U2F: [{ KeyHandle: 'SomeKeyHandle', Label: 'A label' }]
                    }
                }
            };

            expect(render(<ScopeFormModal scope={scope}/>)).toMatchSnapshot();
        });

        test('with U2F and TOTP', () => {
            const scope = {
                response: {
                    TwoFactor: 1,
                    '2FA': {
                        U2F: [{ KeyHandle: 'SomeKeyHandle', Label: 'A label' }],
                        TOTP: 1
                    }
                }
            };

            expect(render(<ScopeFormModal scope={scope}/>)).toMatchSnapshot();
        });

        test('with U2F resolved', () => {
            const scope = {
                response: {
                    TwoFactor: 1,
                    '2FA': {
                        U2F: [{ KeyHandle: 'SomeKeyHandle', Label: 'A label' }],
                        TOTP: 0
                    }
                },
                creds: {
                    U2F: {
                        ClientData: 'the body of U2F is not full but it\'s enough for this test'
                    }
                }
            };

            expect(render(<ScopeFormModal scope={scope}/>)).toMatchSnapshot();
        });

        test('with U2F pending', () => {
            const scope = {
                response: {
                    TwoFactor: 1,
                    '2FA': {
                        U2F: [{ KeyHandle: 'SomeKeyHandle', Label: 'A label' }],
                        TOTP: 0
                    }
                },
                creds: {},
                U2FRequest: {
                    status: 'pending'
                }
            };

            expect(render(<ScopeFormModal scope={scope}/>)).toMatchSnapshot();
        });

        test('with error', () => {
            const scope = {
                response: {
                    TwoFactor: 1,
                    '2FA': {
                        U2F: [{ KeyHandle: 'SomeKeyHandle', Label: 'A label' }],
                        TOTP: 0
                    }
                },
                creds: {},
                U2FRequest: {
                    status: 'failure',
                    error: {
                        metaData: { code: ERROR_CODE.OTHER_ERROR }
                    }
                }
            };

            expect(render(<ScopeFormModal scope={scope}/>)).toMatchSnapshot();

            scope.U2FRequest.error = new Error('Some random error');
            expect(render(<ScopeFormModal scope={scope}/>)).toMatchSnapshot();
        });

    });

    describe('possible actions', () => {
        test('scope already used: ', () => {
            const skip = jest.fn();
            render(<ScopeFormModal scope={{ used: true }} skip={skip}/>);
            expect(skip).toHaveBeenCalledTimes(1);
        });

        test('componentDidMount correctly reset creds', () => {
            const scope = {
                response: {
                    TwoFactor: 1,
                    '2FA': {
                        U2F: [{ KeyHandle: 'SomeKeyHandle', Label: 'A label' }],
                        TOTP: 0
                    }
                },
                creds: {
                    password: 'some password',
                    twoFactorCode: '123456'
                }
            };

            const unscopeInitAction = jest.fn();

            expect(deep(<ScopeFormModal scope={scope} unscopeInitAction={unscopeInitAction}/>)).toMatchSnapshot();
            expect(unscopeInitAction).toHaveBeenCalledTimes(1);
        });

        test('click on use U2F key', () => {
            const scope = {
                response: {
                    TwoFactor: 1,
                    '2FA': {
                        U2F: [{ KeyHandle: 'SomeKeyHandle', Label: 'A label' }]
                    }
                }
            };

            const actions = {
                unscopeInitAction: jest.fn(),
                unscopeU2FAction: jest.fn()
            };

            const tree = deep(<ScopeFormModal scope={scope} {...actions}/>);
            tree.find('a').first().simulate('click');
            expect(actions.unscopeInitAction).toHaveBeenCalledTimes(1);
            expect(actions.unscopeU2FAction).toHaveBeenCalledTimes(1);
        });

        test('input password and code updated', () => {
            const scope = {
                response: {
                    TwoFactor: 1,
                    '2FA': {
                        U2F: [{ KeyHandle: 'SomeKeyHandle', Label: 'A label' }]
                    }
                }
            };

            const actions = {
                unscopeInitAction: jest.fn()
            };

            const tree = deep(<ScopeFormModal scope={scope} {...actions}/>);
            tree.find('input').first().simulate('input', { target: { name: 'password', value: 'example password' } });
            expect(tree).toMatchSnapshot();

            tree.find('input').at(1).simulate('input', { target: { name: 'twoFactorCode', value: '654321' } });
            expect(tree).toMatchSnapshot();

        });

        test('submit and cancel', () => {
            const event = { preventDefault: () => undefined };

            const scope = {
                response: {
                    TwoFactor: 1,
                    '2FA': {
                        U2F: [{ KeyHandle: 'SomeKeyHandle', Label: 'A label' }]
                    }
                },
                creds: {
                    password: 'some password',
                    twoFactorCode: '123456'
                }
            };

            const actions = {
                unscopeInitAction: jest.fn(),
                unscopePasswordAction: jest.fn(),
                onSubmit: jest.fn(),
                onCancel: jest.fn()
            };

            const tree = deep(<ScopeFormModal scope={scope} {...actions}/>);
            tree.find('form').simulate('submit', event);
            expect(actions.onSubmit).toHaveBeenCalled();
            expect(actions.onCancel).not.toHaveBeenCalled();
            expect(actions.unscopePasswordAction).toHaveBeenCalledTimes(1);
            expect(actions.unscopePasswordAction).toHaveBeenCalledWith(scope.creds);

            tree.find('form').simulate('reset', event);
            expect(actions.onCancel).toHaveBeenCalled(); // only once
            expect(actions.onSubmit).toHaveBeenCalledTimes(1); // the submit event from the first call
            expect(actions.unscopePasswordAction).toHaveBeenCalledTimes(1);

        });
    });
});
