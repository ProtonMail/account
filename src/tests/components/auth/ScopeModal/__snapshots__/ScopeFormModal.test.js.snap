// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScopeModal step ScopeFormModal initial display requesting authorization 1`] = `
<form class="wrapper">
  <div class="content ">
    <div class="scopeFormModal form-row">
      <label htmlFor="password" class="label">password</label>
      <div>
        <input class="input" type="password" name="password" id="password" required
        placeholder="Password" autoFocus />
      </div>
    </div>
    <div class="scopeFormModal">Requesting authorization...</div>
  </div>
  <footer class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit" disabled>Next</button>
  </footer>
</form>
`;

exports[`ScopeModal step ScopeFormModal initial display with TOTP only 1`] = `
<form class="wrapper">
  <div class="content ">
    <div class="scopeFormModal form-row">
      <label htmlFor="password" class="label">password</label>
      <div>
        <input class="input" type="password" name="password" id="password" required
        placeholder="Password" autoFocus />
      </div>
    </div>
    <div class="scopeFormModal form-row">
      <label htmlFor="twoFactorCode" class="label">2FA code</label>
      <div>
        <input class="input" type="text" name="twoFactorCode" id="twoFactorCode"
        autoCapitalize="off" autoCorrect="off" autoComplete="off" minLength="6"
        maxLength="8" required />
      </div>
    </div>
  </div>
  <footer class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit">Next</button>
  </footer>
</form>
`;

exports[`ScopeModal step ScopeFormModal initial display with U2F and TOTP 1`] = `
<form class="wrapper">
  <div class="content ">
    <div class="scopeFormModal form-row">
      <label htmlFor="password" class="label">password</label>
      <div>
        <input class="input" type="password" name="password" id="password" required
        placeholder="Password" autoFocus />
      </div>
    </div>
    <div class="scopeFormModal form-row">
      <label htmlFor="twoFactorCode" class="label">2FA code</label>
      <div>
        <input class="input" type="text" name="twoFactorCode" id="twoFactorCode"
        autoCapitalize="off" autoCorrect="off" autoComplete="off" minLength="6"
        maxLength="8" required />
      </div>
    </div>
    <div class="scopeFormModal">
      <p>
        <a class="a" href="#">Or use your security key</a>
      </p>
    </div>
  </div>
  <footer class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit">Next</button>
  </footer>
</form>
`;

exports[`ScopeModal step ScopeFormModal initial display with U2F only 1`] = `
<form class="wrapper">
  <div class="content ">
    <div class="scopeFormModal form-row">
      <label htmlFor="password" class="label">password</label>
      <div>
        <input class="input" type="password" name="password" id="password" required
        placeholder="Password" autoFocus />
      </div>
    </div>
    <div class="scopeFormModal form-row">
      <label htmlFor="twoFactorCode" class="label">2FA code</label>
      <div>
        <input class="input" type="text" name="twoFactorCode" id="twoFactorCode"
        autoCapitalize="off" autoCorrect="off" autoComplete="off" minLength="6"
        maxLength="8" required />
      </div>
    </div>
    <div class="scopeFormModal">
      <p>
        <a class="a" href="#">Or use your security key</a>
      </p>
    </div>
  </div>
  <footer class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit">Next</button>
  </footer>
</form>
`;

exports[`ScopeModal step ScopeFormModal initial display with U2F pending 1`] = `
<form class="wrapper">
  <div class="content ">
    <div class="scopeFormModal form-row">
      <label htmlFor="password" class="label">password</label>
      <div>
        <input class="input" type="password" name="password" id="password" required
        placeholder="Password" autoFocus />
      </div>
    </div>
    <div class="scopeFormModal">
      <p>
        <span>Please activate your security key...</span>
        <a class="a" href="#">Cancel</a>
      </p>
    </div>
  </div>
  <footer class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit">Next</button>
  </footer>
</form>
`;

exports[`ScopeModal step ScopeFormModal initial display with U2F resolved 1`] = `
<form class="wrapper">
  <div class="content ">
    <div class="scopeFormModal form-row">
      <label htmlFor="password" class="label">password</label>
      <div>
        <input class="input" type="password" name="password" id="password" required
        placeholder="Password" autoFocus />
      </div>
    </div>
    <div class="scopeFormModal">
      <span>Your security key was used.</span>
      <a class="a" href="#">Undo</a>
    </div>
  </div>
  <footer class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit">Next</button>
  </footer>
</form>
`;

exports[`ScopeModal step ScopeFormModal initial display with a message 1`] = `
<form class="wrapper">
  <div class="content ">
    <div>This is a simple message</div>
    <div class="scopeFormModal form-row">
      <label htmlFor="password" class="label">password</label>
      <div>
        <input class="input" type="password" name="password" id="password" required
        placeholder="Password" autoFocus />
      </div>
    </div>
    <div class="scopeFormModal"></div>
  </div>
  <footer class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit">Next</button>
  </footer>
</form>
`;

exports[`ScopeModal step ScopeFormModal initial display with error 1`] = `
<form class="wrapper">
  <div class="content ">
    <div class="scopeFormModal form-row">
      <label htmlFor="password" class="label">password</label>
      <div>
        <input class="input" type="password" name="password" id="password" required
        placeholder="Password" autoFocus />
      </div>
    </div>
    <div class="scopeFormModal">
      <p>
        <span>An error occurred You can</span>
        <a class="a" href="#">Try again</a>
        <span>or</span>
        <a class="a" href="#">Use a code</a>
        <span>.</span>
      </p>
    </div>
  </div>
  <footer class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit">Next</button>
  </footer>
</form>
`;

exports[`ScopeModal step ScopeFormModal initial display with error 2`] = `
<form class="wrapper">
  <div class="content ">
    <div class="scopeFormModal form-row">
      <label htmlFor="password" class="label">password</label>
      <div>
        <input class="input" type="password" name="password" id="password" required
        placeholder="Password" autoFocus />
      </div>
    </div>
    <div class="scopeFormModal">
      <p>
        <span>Some random error. You can</span>
        <a class="a" href="#">Try again</a>
        <span>or</span>
        <a class="a" href="#">Use a code</a>
        <span>.</span>
      </p>
    </div>
  </div>
  <footer class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit">Next</button>
  </footer>
</form>
`;

exports[`ScopeModal step ScopeFormModal initial display without 2FA 1`] = `
<form class="wrapper">
  <div class="content ">
    <div class="scopeFormModal form-row">
      <label htmlFor="password" class="label">password</label>
      <div>
        <input class="input" type="password" name="password" id="password" required
        placeholder="Password" autoFocus />
      </div>
    </div>
    <div class="scopeFormModal"></div>
  </div>
  <footer class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit">Next</button>
  </footer>
</form>
`;

exports[`ScopeModal step ScopeFormModal possible actions componentDidMount correctly reset creds 1`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content ">
    <div class="scopeFormModal form-row">
      <label
        htmlFor="password"
        class="label"
      >
        password
      </label>
      <div>
        <input
          class="input"
          type="password"
          name="password"
          id="password"
          value="some password"
          required={true}
          placeholder="Password"
          onInput={[Function onInput]}
          autoFocus={true}
         />
      </div>
    </div>
    <div class="scopeFormModal form-row">
      <label
        htmlFor="twoFactorCode"
        class="label"
      >
        2FA code
      </label>
      <div>
        <input
          class="input"
          type="text"
          name="twoFactorCode"
          id="twoFactorCode"
          autoCapitalize="off"
          autoCorrect="off"
          autoComplete="off"
          minLength="6"
          maxLength="8"
          required={true}
          value="123456"
          onInput={[Function onInput]}
         />
      </div>
    </div>
    <div class="scopeFormModal">
      <p>
        <a
          onClick={[Function onClick]}
          class="a"
          href="#"
        >
          Or use your security key
        </a>
      </p>
    </div>
  </div>
  <footer class="footer">
    <button
      type="reset"
      value="Reset"
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
      disabled={false}
    >
      Next
    </button>
  </footer>
</form>

`;

exports[`ScopeModal step ScopeFormModal possible actions input password and code updated 1`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content ">
    <div class="scopeFormModal form-row">
      <label
        htmlFor="password"
        class="label"
      >
        password
      </label>
      <div>
        <input
          class="input"
          type="password"
          name="password"
          id="password"
          value="example password"
          required={true}
          placeholder="Password"
          onInput={[Function onInput]}
          autoFocus={true}
         />
      </div>
    </div>
    <div class="scopeFormModal form-row">
      <label
        htmlFor="twoFactorCode"
        class="label"
      >
        2FA code
      </label>
      <div>
        <input
          class="input"
          type="text"
          name="twoFactorCode"
          id="twoFactorCode"
          autoCapitalize="off"
          autoCorrect="off"
          autoComplete="off"
          minLength="6"
          maxLength="8"
          required={true}
          onInput={[Function onInput]}
         />
      </div>
    </div>
    <div class="scopeFormModal">
      <p>
        <a
          onClick={[Function onClick]}
          class="a"
          href="#"
        >
          Or use your security key
        </a>
      </p>
    </div>
  </div>
  <footer class="footer">
    <button
      type="reset"
      value="Reset"
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
      disabled={false}
    >
      Next
    </button>
  </footer>
</form>

`;

exports[`ScopeModal step ScopeFormModal possible actions input password and code updated 2`] = `
preact-render-spy (1 nodes)
-------
<form
  class="wrapper"
  onSubmit={[Function onSubmit]}
  onReset={[Function onReset]}
>
  <div class="content ">
    <div class="scopeFormModal form-row">
      <label
        htmlFor="password"
        class="label"
      >
        password
      </label>
      <div>
        <input
          class="input"
          type="password"
          name="password"
          id="password"
          value="example password"
          required={true}
          placeholder="Password"
          onInput={[Function onInput]}
          autoFocus={true}
         />
      </div>
    </div>
    <div class="scopeFormModal form-row">
      <label
        htmlFor="twoFactorCode"
        class="label"
      >
        2FA code
      </label>
      <div>
        <input
          class="input"
          type="text"
          name="twoFactorCode"
          id="twoFactorCode"
          autoCapitalize="off"
          autoCorrect="off"
          autoComplete="off"
          minLength="6"
          maxLength="8"
          required={true}
          value="654321"
          onInput={[Function onInput]}
         />
      </div>
    </div>
    <div class="scopeFormModal">
      <p>
        <a
          onClick={[Function onClick]}
          class="a"
          href="#"
        >
          Or use your security key
        </a>
      </p>
    </div>
  </div>
  <footer class="footer">
    <button
      type="reset"
      value="Reset"
    >
      Back
    </button>
    <button
      type="submit"
      value="Submit"
      disabled={false}
    >
      Next
    </button>
  </footer>
</form>

`;
